Firestore/core/src/firebase/firestore/util/filesystem_posix.cc

```diff
@@ -130,51 +131,88 @@ Status DeleteFile(const Path& path) {
 }
 
 Status RecursivelyDeleteDir(const Path& parent) {
-  DIR* dir = ::opendir(parent.c_str());
-  if (!dir) {
-    if (errno == ENOENT) {
-      return Status::OK();
+  DirectoryIterator iter(parent);
+  for (; iter.Valid(); iter.Next()) {
+    Status status = RecursivelyDelete(iter.file());
+    if (!status.ok()) {
+      return status;
     }
-
-    return Status::FromErrno(errno, StringFormat("Could not read directory %s",
-                                                 parent.ToUtf8String()));
   }
+  if (!iter.status().ok()) {
+    return iter.status();
+  }
+  return DeleteDir(parent);
+}
 
-  Status result;
-  while (result.ok()) {
-    errno = 0;
-    struct dirent* entry = ::readdir(dir);
-    if (!entry) {
-      if (errno != 0) {
-        result = Status::FromErrno(
-            errno,
-            StringFormat("Could not read directory %s", parent.ToUtf8String()));
-      }
-      break;
-    }
+}  // namespace detail
 
-    if (::strcmp(".", entry->d_name) == 0 ||
-        ::strcmp("..", entry->d_name) == 0) {
-      continue;
-    }
+struct DirectoryIterator::Rep {
+  DIR* dir;
+  struct dirent* entry;
+};
 
-    Path child = parent.AppendUtf8(entry->d_name, strlen(entry->d_name));
-    result = RecursivelyDelete(child);
+DirectoryIterator::DirectoryIterator(
+    const firebase::firestore::util::Path& path)
+    : parent_(path), rep_(absl::make_unique<DirectoryIterator::Rep>()) {
```

Comment:
nit: we've been using brace-style initialization when possible, which would make these `parent_{path}`, etc.